src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs

```diff
@@ -868,5 +869,80 @@ public void ReusingNames_Windows(string name)
             }
         }
 
+        private void ValidateDeviceAccess(MemoryMappedFile memMap, long viewCapacity, MemoryMappedFileAccess access)
+        {
+            using (MemoryMappedViewAccessor view = memMap.CreateViewAccessor(0, viewCapacity, access))
+            {
+                if (access != MemoryMappedFileAccess.Write)
+                {
+                    byte b = view.ReadByte(0);
+                    // /dev/zero return zeroes.
+                    Assert.Equal(0, b);
+                }
+
+                if (access != MemoryMappedFileAccess.Read)
+                {
+                    view.Write(0, (byte)1);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Test that we can map special character devices on Unix using FileStream.
+        /// </summary>
+        [ConditionalTheory]
+        [InlineData(MemoryMappedFileAccess.Read)]
+        [InlineData(MemoryMappedFileAccess.ReadWrite)]
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        public void OpenCharacterDeviceAsStream(MemoryMappedFileAccess access)
+        {
+            string device = "/dev/zero";
+            if (!File.Exists(device))
+            {
+                throw new SkipTestException($"'{device}' is not available.");
+            }
+
+            long viewCapacity = 0xFF;
+
+            try
+            {
+                using (FileStream fs = new FileStream(device, FileMode.Open, FileAccess.ReadWrite, FileShare.ReadWrite))
+                using (MemoryMappedFile memMap = MemoryMappedFile.CreateFromFile(fs, null, viewCapacity, access, HandleInheritability.None, false))
+                {
+                    ValidateDeviceAccess(memMap, viewCapacity, access);
+                }
+            }
+            catch (UnauthorizedAccessException) { }
+            // ENODEV Operation not supported by device.
```

Comment:
The ENODEV is coming from OSX. I did not test on FreeBSD.  I could make the test Platform specific. But since the code is now unix specific the test verifies that at least we do not throw some unexpected exception.